### ISSUE TITLE
-Wreturn-type should not emit warnings when non-termination can be derived indirectly

### ISSUE THREAD

**Issue by EdSchouten** [llvm/llvm-project#11914]
|  |  |
| --- | --- |
| Bugzilla Link | [11542](https://llvm.org/bz11542) |
| Version | 3.0 |
| OS | FreeBSD |
| CC | @efriedma-quic,@tkremenek |

## Extended Description 
I suspect this bug is similar to bug 9599, but not identical, so I'm filing this anyway.

The following code generates a -Wreturn-type warning:
```c
__attribute__((__noreturn__)) void bar(void);

static void
foo(void)
{

        bar();
}

int
f(void)
{

        foo();
}
```

foo.c:15:1: warning: control reaches end of non-void function [-Wreturn-type]

If I change the call to foo() to bar(), it obviously work, but in this case the compiler should be to derive to a certain extent that foo() isn't going to return anyway. This information should be available anyway, as for this specific C file, foo() is even inlined into bar() completely.

**Comment by philnik777** [comment]
This seems to me more like we should enable `-Wmissing-noreturn` by default to propagate `[[noreturn]]`, or at least with `-Wall` or `-Wextra`.

**Comment by AaronBallman** [comment]
Resolving as a duplicate of #111509
